storage/storage.go

```diff
@@ -90,55 +90,131 @@ func (storage *Storage) getAllChunks(context context.Context, buildId string) ([
+	if first.Before(second) {
+		return first
+	} else {
+		return second
 	}
+}
 
-	testChunkIterator := NewBatchedLogIterator(storage.bucket, testChunks, 4, testTimeRange)
+func maxTime(first time.Time, second time.Time) time.Time {
+	if first.Before(second) {
+		return second
+	} else {
+		return first
+	}
+}
 
-	buildChunks := []LogChunkInfo{}
+func (storage *Storage) GetAllLogLines(context context.Context, buildId string) (LogIterator, error) {
+	buildChunks, testChunks, err := storage.getBuildAndTestChunks(context, buildId)
+	if err != nil {
+		return nil, err
+	}
+
+	sortByStartTime(buildChunks)
+	sortByStartTime(testChunks)
+
+	timeRange := TimeRange{
+		StartAt: minTime(buildChunks[0].Start, testChunks[0].Start),
+		EndAt:   maxTime(getLatestTime(buildChunks), getLatestTime(testChunks)),
+	}
+
+	buildChunkIterator := NewBatchedLogIterator(storage.bucket, buildChunks, 4, timeRange)
+	testChunkIterator := NewBatchedLogIterator(storage.bucket, testChunks, 4, timeRange)
+
+	// Merge test and build logs
+	return NewMergingIterator(testChunkIterator, buildChunkIterator), nil
+}
+
+func (storage *Storage) GetGlobalLogLines(context context.Context, buildId string) (LogIterator, error) {
+	buildChunks, _, err := storage.getBuildAndTestChunks(context, buildId)
+	if err != nil {
+		return nil, err
+	}
+
+	sortByStartTime(buildChunks)
+
+	timeRange := TimeRange{
+		StartAt: buildChunks[0].Start,
+		EndAt:   getLatestTime(buildChunks),
+	}
+
+	buildChunkIterator := NewBatchedLogIterator(storage.bucket, buildChunks, 4, timeRange)
+
+	// Merge everything together
+	return buildChunkIterator, nil
+}
+
+func testChunksWithId(chunks []LogChunkInfo, testID string) []LogChunkInfo {
+	testChunks := []LogChunkInfo{}
 	for i := 0; i < len(chunks); i++ {
-		// Include any build logs that are in the time range of our test
-		chunkTimeRange := TimeRange{
-			StartAt: chunks[i].Start,
-			EndAt:   chunks[i].End,
-		}
-		// check if the global build chunk's time range intersects the test's time range, and if so
-		// add it to our list of build chunks, but constrained to the test's time range to only
-		// include entries during that time.
-		if chunks[i].TestID == "" && testTimeRange.Intersects(chunkTimeRange) {
-			buildChunks = append(buildChunks, chunks[i])
+		// Find our test id
+		if chunks[i].TestID == testID {
+			testChunks = append(testChunks, chunks[i])
 		}
 	}
+	return testChunks
+}
 
-	sort.Slice(buildChunks, func(i, j int) bool {
-		return buildChunks[i].Start.Before(buildChunks[j].Start)
-	})
+func (storage *Storage) GetTestLogLines(context context.Context, buildId string, testId string) (LogIterator, error) {
+	buildChunks, allTestChunks, err := storage.getBuildAndTestChunks(context, buildId)
+	if err != nil {
+		return nil, err
+	}
+
+	testChunks := testChunksWithId(allTestChunks, testId)
+
+	sortByStartTime(testChunks)
+
+	testTimeRange := TimeRange{
+		StartAt: testChunks[0].Start,
+		EndAt:   getLatestTime(testChunks),
```

Comment:
It seems like this could be different from the way the end time is currently determined. There the maxTime [is the start of the next test](https://github.com/evergreen-ci/logkeeper/blob/5957d4855c595454db0fc288df6ac2d2b6fb9e46/views.go#L625).
The difference would show if there's a pause between tests but there are ongoing global logs: today logkeeper returns global logs from the interim, whereas determining the end of the time range from the last test chunk will preclude them.

Reply:
This should be fixed. We technically get the beginning of whatever test chunk has a start time bigger than the end time of the latest chunk of the test. I think that essentially does the same thing assuming the tests are non-overlapping. If the tests are overlapping then the new logic would find some test chunk in the other test that was strictly after this test, which seems... fine. Although we are assuming that the tests are non-overlapping.